src/vm/docs/mod.rs

```diff
@@ -0,0 +1,426 @@
+principal, and executes `expr` with that context. It returns the resulting value of `expr`.",
+    example: "(as-contract (print tx-sender)) => 'CTcontract.name"
+};
+
+
+fn make_for_special(api: &SpecialAPI) -> FunctionAPI {
+    FunctionAPI {
+        input_type: api.input_type.to_string(),
+        output_type: api.output_type.to_string(),
+        signature: api.signature.to_string(),
+        description: api.description.to_string(),
+        example: api.example.to_string()
+    }
+}
+
+
+fn make_api_reference(function: &NativeFunctions) -> FunctionAPI {
+    use vm::functions::NativeFunctions::*;
+    match function {
+        Add => make_for_simple_native(&ADD_API, &Add),
+        Subtract => make_for_simple_native(&SUB_API, &Subtract),
+        Multiply => make_for_simple_native(&MUL_API, &Multiply),
+        Divide => make_for_simple_native(&DIV_API, &Divide),
+        CmpGeq => make_for_simple_native(&GEQ_API, &CmpGeq),
+        CmpLeq => make_for_simple_native(&LEQ_API, &CmpLeq),
+        CmpLess => make_for_simple_native(&LESS_API, &CmpLess),
+        CmpGreater => make_for_simple_native(&GREATER_API, &CmpGreater),
+        Modulo => make_for_simple_native(&MOD_API, &Modulo),
+        Power => make_for_simple_native(&POW_API, &Power),
+        BitwiseXOR => make_for_simple_native(&XOR_API, &BitwiseXOR),
+        And => make_for_simple_native(&AND_API, &And),
+        Or => make_for_simple_native(&OR_API, &Or),
+        Not => make_for_simple_native(&NOT_API, &Not),
+        Equals => make_for_simple_native(&EQUALS_API, &Equals),
+        If => make_for_special(&IF_API),
+        Let => make_for_special(&LET_API),
+        Map => make_for_special(&MAP_API),
+        Fold => make_for_special(&FOLD_API),
+        ListCons => make_for_special(&LIST_API),
+        FetchEntry => make_for_special(&FETCH_API),
+        FetchContractEntry => make_for_special(&FETCH_CONTRACT_API),
+        SetEntry => make_for_special(&SET_API),
+        InsertEntry => make_for_special(&INSERT_API),
+        DeleteEntry => make_for_special(&DELETE_API),
+        TupleCons => make_for_special(&TUPLE_CONS_API),
+        TupleGet => make_for_special(&TUPLE_GET_API),
+        Begin => make_for_special(&BEGIN_API),
+        Hash160 => make_for_special(&HASH_160_API),
+        Print => make_for_special(&PRINT_API),
+        ContractCall => make_for_special(&CONTRACT_CALL_API),
+        AsContract => make_for_special(&AS_CONTRACT_API)
+    }
+}
+
+pub fn make_json_api_reference() -> String {
+    use vm::functions::NativeFunctions::*;
+    let natives = [ Add, Subtract, Multiply, Divide, CmpGeq, CmpLeq, CmpLess, CmpGreater, Modulo, Power,
+                    BitwiseXOR, And, Or, Not, Equals ];
+    let json_references: Vec<_> = natives.iter()
+        .map(|x| make_api_reference(x))
+        .collect();
+    format!("{}", serde_json::to_string(&json_references)
+            .expect("Failed to serialize documentation"))
+}
```

Comment:
I appreciate what's going on here, but is there a "standard" way of doing this with Rust documentation tools?  Or possibly using a combination of Rust documentation tools and scripts we supply?

Reply:
Not that I could tell -- rust's standard doc generation fails on a number of accounts. First, it documents the _rust_ functions themselves, which these aren't. Second, it only outputs HTML, and for convenience for integrating with other documentation, we want JSON. Third, there are a bunch of fields _we're_ interested in that rustdoc just isn't.

Reply:
Okay, sounds good.